### DISCUSSION TITLE
How to add the vector to a field with vector_index

### DISCUSSION THREAD

**Discussion by ujjwalm29** [neo4j-contrib/neomodel#823] [Q&A] [answered]
Hello @mariusconjeaud,

Been loving everything about neomodel so far. I have a question about the `vector_index` field in the `StringProperty()`.

How do you actually add the vector to the index?

For a `fulltext_index`, I am assuming that when I do `node.string_field = "data"`, the indexing on this data is triggered on it's own.
How does it work for `vector_index`? How do I enter my vector? Is it something like `node.string_field.vector=[...]`?

Thanks for your help. I am specifically tagging you since [this](https://github.com/neo4j-contrib/neomodel/pull/815) is the only best available documentation related to vector_index. I went through the code but couldn't find anything. If I missed something, please tell me where I can find it. Thank you!

**Comment by mariusconjeaud** [comment, marked as answer]
Hello,

Firstly, I added a section specifically about Schema in the documentation. There was an issue in the docs generation at release time though but it was fixed some time back.
Have a look here : https://neomodel.readthedocs.io/en/latest/schema_management.html

Then, about "adding a vector to the index", I am not 100% sure I got your question right. If I read correctly, you are interested in knowing how actual data fields are added into the index inside of Neo4j ? This happens automatically once the property is saved to the database and is handled by Neo4j itself.
So that when you do : `myNode.vector = [1.0, 1.0]` and then `myNode.save()`, the index inside of Neo4j will be updated since the value of an indexed property changed on a node.

**Reply by ujjwalm29** [reply]
Ok so I was looking for `myNode.vector = [1.0, 1.0]`.

I am still confused about whether the vector is a separate property of the node or it's **attached** to the specific data field.


Let's say I have a class : 
```
class MyNode(StructuredNode):
    name = StringProperty(fulltext_index=FulltextIndex(), vector_index=VectorIndex(dimensions=DIMENSIONS))
    description = StringProperty(fulltext_index=FulltextIndex(), vector_index=VectorIndex(dimensions=DIMENSIONS))
```

Then, to create a node, I do : 
```
myNode = MyNode(name="myName", description="awesomeDescription").save()
myNode.vector = ?
```

Questions : 
- Are multiple properties in a single node allowed to have their own separate vector indexes? The neo4j vector index doc says _"A vector index is a single-label, single-property index for nodes or a single-relationship-type, single-property index for relationships. "_ . But can 2 separate indexes be created on 2 separate properties in a single node? If yes, how do I assign their vectors? Because. now, `myNode.vector=[1.0,1.0]` is ambiguous.

- Just for confirmation, do I need to do any extra step to trigger the full text indexing ? Or is it enough to just do `MyNode(name="myName", description="awesomeDescription").save()` ?

**Reply by mariusconjeaud** [reply]
> Are multiple properties in a single node allowed to have their own separate vector indexes

Yes, your MyNode class definition is fine.

> do I need to do any extra step to trigger the full text indexing

No, the moment you save a property to the database, the indexing mechanism of Neo4j will do its work.
So, both of these work :

```python
# Create the node and save it
my_node = MyNode(name="myName", description="awesomeDescription").save()

# Update the node and save it
my_node.description = "A different but cooler description"
my_node.save()
```

**Reply by ujjwalm29** [reply]
Great thanks, this has been very helpful!

Any word on how do I insert the vector?

```
class MyNode(StructuredNode):
    name = StringProperty(fulltext_index=FulltextIndex(), vector_index=VectorIndex(dimensions=DIMENSIONS))
    description = StringProperty(fulltext_index=FulltextIndex(), vector_index=VectorIndex(dimensions=DIMENSIONS))
    
my_node = MyNode(name="myName", description="awesomeDescription").save()

name_vector: List[float] = get_vector(name)
description_vector: List[float] = get_vector(description)
```

How do i insert these vectors in my respective indexes :

Is it 
```
my_node.name.vector = name_vector
my_node.description.vector = description_vector
```
?

**Reply by mariusconjeaud** [reply]
Ahhhhh, I finally understood what was your question exactly. A Vector index in Neo4j, is just a special **type** of index, which applies to arrays of floats, and enables computing the similarity between pairs of entries.

So, you can only create vector indexes for the ArrayProperty(FloatProperty()) type in neomodel, like :

```python
class MyNode(StructuredNode):
    name = StringProperty(fulltext_index=FulltextIndex())
    description = StringProperty(fulltext_index=FulltextIndex())
    description_embedding = ArrayProperty(FloatProperty(), vector_index=VectorIndex(dimensions=DIMENSIONS))
```

And I realize that the current implementation of neomodel on this is potentially wrong, and the documentation for sure is wrong because we are doing as in your example and applying the Vector index to a StringProperty, which is stupid => vector = array of floats.

So in short : if you want fulltext index + vector index for the same source property (a piece of text), you need two properties : one StringProperty with fulltext_index, and one ArrayProperty(FloatProperty()) with vector_index

**Reply by ujjwalm29** [reply]
Hey @mariusconjeaud thank you very much for clarifying this! Do you want me to create a github issue for this(and potentially fix it?) or is this being tracked somewhere?

**Reply by mariusconjeaud** [reply]
I fixed the documentation and tests already, and it is merge in the release candidate branch rc/5.3.3
